File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-4.cpp
class Solution {
public:
    int characterReplacement(string s, int k) {
        vector<int> count(26, 0);
        int maxFreq = 0;
        int left = 0;
        int maxLen = 0;

        for (int right = 0; right < (int)s.size(); right++) {
            int idx = s[right] - 'A';
            count[idx]++;
            maxFreq = max(maxFreq, count[idx]);

            while ((right - left + 1) - maxFreq > k) {
                count[s[left] - 'A']--;
                left++;
            }

            maxLen = max(maxLen, right - left + 1);
        }

        return maxLen;
    }
};

File: Data Structures & Algorithms/binary-tree-diameter/submission-4.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number}
     */
    diameterOfBinaryTree(root: TreeNode | null): number {
        let diameter = 0;

        const height = (node: TreeNode | null): number => {
            if (node === null) return 0;

            const left = height(node.left);
            const right = height(node.right);

            diameter = Math.max(diameter, left + right);

            return Math.max(left, right) + 1;
        };

        height(root);
        return diameter;
    }
}